[SR-3871] Unwrap multiple layers of Any/AnyObject when casting to protocol type

tbkka opened the pull request:
This fixes an oversight in the runtime dynamic cast operator:  When casting to a protocol-constrained existential type, it would unwrap at most one layer of existential containers.

As a result, code that mixed Any with AnyObject and/or Obj-C passing would be unable to cast the result to a protocol type.

Resolves SR-3871 (and some but likely not all of the duplicates)

tbkka committed:
SR-3871: Dynamic casting of existentials stored in Obj-C references

Arbitrary Swift objects get packaged into __SwiftValue containers so
that pointers to them can be passed into Obj-C.  (Obviously, Obj-C
code can't do anything particularly useful with such pointers other
than refcount them and give them back to Swift code.)  Those values come
back into Swift as either `Any` (existential box) or `AnyObject`
(anonymous object pointer) values.  Dynamically casting those requires
first inspecting the outer value to get access to the actual type and
value in the __SwiftValue container.

The tryDynamicCastBoxedSwiftValue() function that handles this
was missing a check for the `Any` case, which is why directly
casting from `Any` would routinely fail.

tbkka committed:
Minor code reformatting

tbkka committed:
Merge branch 'master' into SR-3871

tbkka committed:
Merge branch 'master' into SR-3871

tbkka committed:
Merge branch 'master' into SR-3871

tbkka committed:
Fill in some additional test assertions

tbkka commented:
@swift-ci Please test Linux

tbkka commented:
@swift-ci Please test macOS

tbkka committed:
Disable Obj-C test on Linux

tbkka commented:
@swift-ci Please test Linux

tbkka committed:
Don't have Linux tests rely on protocols defined for mac-only tests

tbkka commented:
@swift-ci Please test Linux

tbkka commented:
@swift-ci Please smoke test macOS

tbkka committed:
Merge branch 'master' into SR-3871

slavapestov commented on stdlib/public/runtime/Casting.cpp:
```diff
@@ -1879,11 +1879,34 @@ static bool tryDynamicCastBoxedSwiftValue(OpaqueValue *dest,
+      // If it's an opaque existential, unwrap it and check again
+      auto opaqueContainer = reinterpret_cast<OpaqueExistentialContainer*>(src);
+      srcType = opaqueContainer->Type;
+      src = existentialType->projectValue(src);
```
If `flags & DynamicCastFlags::TakeOnSuccess` is true, do we destroy the innermost `src` or the original `src`? It appears to be doing the former, which I suspect is incorrect.

tbkka committed:
Correctly destroy the original outer object

As pointed out in review comments by Slava, the previous code reused the `src`
and `srcType` variables as it unwrapped boxes and then ultimately released
whatever those pointed to.  Now I keep the original values around as
`originalSrc` and `originalSrcType` and use those in the cleanup code.

Here's how the flags are handled in this function now:
	* TakeOnSuccess always destroys the outer original object using the new `originalSrc*` variables.
  * DestroyOnFailure is never passed into this code, so can be ignored.
	* Unconditional is never passed into this code, so can be ignored.

tbkka committed:
Merge branch 'SR-3871' of github.com:tbkka/swift into SR-3871

tbkka committed:
Merge branch 'master' into SR-3871

tbkka commented:
@swift-ci Please test Linux

tbkka commented:
@swift-ci Please smoke test macOS
